File: Sources/LoomCore/Patterns.swift
import Foundation

public enum LoomPatternStatus: String, Codable, Sendable {
  case draft
  case stable
  case deprecated
}

public enum LoomPatternValueType: String, Codable, Sendable, Equatable {
  case string
  case number
  case boolean
  case enumeration = "enum"
  case length
  case insets
  case alignment
  case binding
  case resource
  case reference
}

public struct LoomPatternIntent: Codable, Sendable {
  public var summary: String
  public var useWhen: [String]
  public var avoidWhen: [String]
}

public struct LoomPatternSemantics: Codable, Sendable {
  public var role: String
  public var childPolicy: String
  public var sizing: String
  public var ordering: String
}

public struct LoomPatternAttribute: Codable, Sendable {
  public var name: String
  public var valueType: LoomPatternValueType
  public var required: Bool
  public var description: String
  public var defaultValue: String?
  public var allowedValues: [String]?
  public var minimum: Double?
  public var maximum: Double?
  public var units: [String]?
}

public struct LoomPatternConstraint: Codable, Sendable {
  public var expression: String
  public var message: String
}

public struct LoomPatternAccessibility: Codable, Sendable {
  public var role: String
  public var nameSource: String
  public var focusBehavior: String
  public var notes: [String]
}

public struct LoomPatternMapping: Codable, Sendable {
  public var platform: String
  public var constructs: [String]
  public var strategy: String
  public var caveats: [String]
}

public struct LoomPattern: Codable, Sendable {
  public var schemaVersion: String
  public var id: String
  public var version: String
  public var name: String
  public var kind: LoomNodeKind
  public var status: LoomPatternStatus
  public var category: String
  public var intent: LoomPatternIntent
  public var semantics: LoomPatternSemantics
  public var attributes: [LoomPatternAttribute]
  public var constraints: [LoomPatternConstraint]
  public var accessibility: LoomPatternAccessibility
  public var mappings: [LoomPatternMapping]
  public var tags: [String]

  enum CodingKeys: String, CodingKey {
    case schemaVersion = "schema_version"
    case id, version, name, kind, status, category, intent, semantics, attributes, constraints
    case accessibility, mappings, tags
  }
}

public struct LoomPatternIssue: Codable, Sendable {
  public var severity: LoomDiagnosticSeverity
  public var code: String
  public var path: String
  public var detail: String
}

public struct LoomPatternValidationReport: Codable, Sendable {
  public var schemaVersion = "1"
  public var status: String
  public var directory: String
  public var patternCount: Int
  public var issues: [LoomPatternIssue]

  enum CodingKeys: String, CodingKey {
    case schemaVersion = "schema_version"
    case status, directory, patternCount, issues
  }
}

public struct LoomPatternCatalog: Sendable {
  public init() {}

  public func load(directory: String) throws -> [LoomPattern] {
    let url = URL(fileURLWithPath: directory).standardizedFileURL
    let files: [URL]
    do {
      files = try FileManager.default.contentsOfDirectory(
        at: url, includingPropertiesForKeys: nil
      ).filter { $0.lastPathComponent.hasSuffix(".pattern.json") }.sorted {
        $0.lastPathComponent < $1.lastPathComponent
      }
    } catch {
      throw LoomError.invalidPattern("Could not read pattern directory at \(url.path).")
    }
    return try files.map { file in
      do {
        return try JSONDecoder().decode(LoomPattern.self, from: Data(contentsOf: file))
      } catch {
        throw LoomError.invalidPattern("\(file.lastPathComponent): \(error.localizedDescription)")
      }
    }
  }

  public func find(_ id: String, directory: String) throws -> LoomPattern? {
    try load(directory: directory).first { $0.id == id }
  }

  public func validate(directory: String) -> LoomPatternValidationReport {
    let standardized = URL(fileURLWithPath: directory).standardizedFileURL.path
    let files: [URL]
    do {
      files = try FileManager.default.contentsOfDirectory(
        at: URL(fileURLWithPath: standardized), includingPropertiesForKeys: nil
      ).filter { $0.lastPathComponent.hasSuffix(".pattern.json") }.sorted {
        $0.lastPathComponent < $1.lastPathComponent
      }
    } catch {
      return report(
        directory: standardized, count: 0,
        issues: [
          issue("PATTERN001", standardized, "Pattern directory cannot be read.")
        ])
    }

    var patterns: [(URL, LoomPattern)] = []
    var issues: [LoomPatternIssue] = []
    for file in files {
      do {
        patterns.append(
          (file, try JSONDecoder().decode(LoomPattern.self, from: Data(contentsOf: file))))
      } catch {
        issues.append(issue("PATTERN002", file.lastPathComponent, error.localizedDescription))
      }
    }
    if files.isEmpty {
      issues.append(issue("PATTERN003", standardized, "No .pattern.json files were found."))
    }

    var ids = Set<String>()
    var kinds = Set<LoomNodeKind>()
    for (file, pattern) in patterns {
      let path = file.lastPathComponent
      if pattern.schemaVersion != "1" {
        issues.append(issue("PATTERN004", path, "schema_version must be 1."))
      }
      if pattern.id.isEmpty
        || !pattern.id.allSatisfy({ $0.isLowercase || $0.isNumber || $0 == "-" })
      {
        issues.append(
          issue("PATTERN005", path, "id must use lowercase letters, numbers, and hyphens."))
      }
      if file.deletingPathExtension().deletingPathExtension().lastPathComponent != pattern.id {
        issues.append(issue("PATTERN006", path, "Filename must be \(pattern.id).pattern.json."))
      }
      if !ids.insert(pattern.id).inserted {
        issues.append(issue("PATTERN007", path, "Duplicate pattern id \(pattern.id)."))
      }
      if !kinds.insert(pattern.kind).inserted {
        issues.append(
          issue("PATTERN008", path, "Duplicate semantic kind \(pattern.kind.rawValue)."))
      }
      if pattern.intent.summary.isEmpty || pattern.semantics.role.isEmpty
        || pattern.category.isEmpty
      {
        issues.append(
          issue("PATTERN009", path, "Intent, semantic role, and category must be non-empty."))
      }
      var attributeNames = Set<String>()
      for attribute in pattern.attributes {
        let attributePath = "\(path)#attributes.\(attribute.name)"
        if attribute.name.isEmpty || attribute.description.isEmpty {
          issues.append(
            issue("PATTERN010", attributePath, "Attribute name and description are required."))
        }
        if !attributeNames.insert(attribute.name).inserted {
          issues.append(issue("PATTERN011", attributePath, "Duplicate attribute name."))
        }
        if let minimum = attribute.minimum, let maximum = attribute.maximum, minimum > maximum {
          issues.append(issue("PATTERN012", attributePath, "minimum cannot exceed maximum."))
        }
        if attribute.valueType == .enumeration && (attribute.allowedValues?.isEmpty != false) {
          issues.append(
            issue("PATTERN013", attributePath, "enum attributes require allowedValues."))
        }
        if let value = attribute.defaultValue, let allowed = attribute.allowedValues,
          !allowed.contains(value)
        {
          issues.append(
            issue("PATTERN014", attributePath, "defaultValue must be an allowed value."))
        }
      }
      if pattern.mappings.isEmpty
        || Set(pattern.mappings.map(\.platform)).count != pattern.mappings.count
      {
        issues.append(
          issue("PATTERN015", path, "Mappings must be non-empty and platform names unique."))
      }
    }
    return report(directory: standardized, count: patterns.count, issues: issues)
  }

  public func json<T: Encodable>(_ value: T) throws -> String {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
    return String(decoding: try encoder.encode(value), as: UTF8.self) + "\n"
  }

  public func text(_ report: LoomPatternValidationReport) -> String {
    var lines = [
      "Loom pattern validation", "Status: \(report.status)",
      "Directory: \(report.directory)", "Patterns: \(report.patternCount)",
      "Issues: \(report.issues.count)",
    ]
    lines += report.issues.map { "  [\($0.severity.rawValue)] \($0.code) \($0.path): \($0.detail)" }
    return lines.joined(separator: "\n") + "\n"
  }

  public func listText(_ patterns: [LoomPattern]) -> String {
    let width = patterns.map(\.id.count).max() ?? 0
    return patterns.map {
      "\($0.id + String(repeating: " ", count: max(0, width - $0.id.count)))  \($0.kind.rawValue)  \($0.intent.summary)"
    }.joined(separator: "\n") + (patterns.isEmpty ? "" : "\n")
  }

  private func issue(_ code: String, _ path: String, _ detail: String) -> LoomPatternIssue {
    LoomPatternIssue(severity: .error, code: code, path: path, detail: detail)
  }

  private func report(directory: String, count: Int, issues: [LoomPatternIssue])
    -> LoomPatternValidationReport
  {
    LoomPatternValidationReport(
      status: issues.isEmpty ? "ok" : "error", directory: directory,
      patternCount: count, issues: issues)
  }
}

File: internal/loom/patterns.go
package loom

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type PatternStatus string
type PatternValueType string

const DefaultPatternDirectory = "patterns"

type PatternIntent struct {
	Summary   string   `json:"summary"`
	UseWhen   []string `json:"useWhen"`
	AvoidWhen []string `json:"avoidWhen"`
}

type PatternSemantics struct {
	Role        string `json:"role"`
	ChildPolicy string `json:"childPolicy"`
	Sizing      string `json:"sizing"`
	Ordering    string `json:"ordering"`
}

type PatternAttribute struct {
	Name          string   `json:"name"`
	ValueType     string   `json:"valueType"`
	Required      bool     `json:"required"`
	Description   string   `json:"description"`
	DefaultValue  string   `json:"defaultValue,omitempty"`
	AllowedValues []string `json:"allowedValues,omitempty"`
	Minimum       *float64 `json:"minimum,omitempty"`
	Maximum       *float64 `json:"maximum,omitempty"`
	Units         []string `json:"units,omitempty"`
}

type PatternConstraint struct {
	Expression string `json:"expression"`
	Message    string `json:"message"`
}

type PatternKeyboard struct {
	Activation     []string `json:"activation,omitempty"`
	Navigation     []string `json:"navigation,omitempty"`
	EscapeBehavior string   `json:"escape_behavior,omitempty"`
}

type PatternMinimumTargetSize struct {
	Width  float64 `json:"width"`
	Height float64 `json:"height"`
	Unit   string  `json:"unit"`
}

type PatternAccessibility struct {
	Role              string                    `json:"role"`
	NameSource        string                    `json:"nameSource"`
	FocusBehavior     string                    `json:"focusBehavior"`
	Notes             []string                  `json:"notes"`
	Keyboard          *PatternKeyboard          `json:"keyboard,omitempty"`
	States            []string                  `json:"states,omitempty"`
	Properties        []string                  `json:"properties,omitempty"`
	MinimumTargetSize *PatternMinimumTargetSize `json:"minimumTargetSize,omitempty"`
}

type PatternMapping struct {
	Platform   string   `json:"platform"`
	Constructs []string `json:"constructs"`
	Strategy   string   `json:"strategy"`
	Caveats    []string `json:"caveats"`
}

type PatternVariant struct {
	Name         string   `json:"name"`
	Conditions   []string `json:"conditions"`
	LayoutPolicy string   `json:"layout_policy"`
	Intent       string   `json:"intent"`
}

type Pattern struct {
	SchemaVersion string               `json:"schema_version"`
	ID            string               `json:"id"`
	Version       string               `json:"version"`
	Name          string               `json:"name"`
	Kind          NodeKind             `json:"kind"`
	Status        PatternStatus        `json:"status"`
	Category      string               `json:"category"`
	Intent        PatternIntent        `json:"intent"`
	Semantics     PatternSemantics     `json:"semantics"`
	Attributes    []PatternAttribute   `json:"attributes"`
	Constraints   []PatternConstraint  `json:"constraints"`
	Accessibility PatternAccessibility `json:"accessibility"`
	Mappings      []PatternMapping     `json:"mappings"`
	Variants      []PatternVariant     `json:"variants,omitempty"`
	Tags          []string             `json:"tags"`
}

type PatternIssue struct {
	Severity DiagnosticSeverity `json:"severity"`
	Code     string             `json:"code"`
	Path     string             `json:"path"`
	Detail   string             `json:"detail"`
}

type PatternValidationReport struct {
	SchemaVersion string         `json:"schema_version"`
	Status        string         `json:"status"`
	Directory     string         `json:"directory"`
	PatternCount  int            `json:"patternCount"`
	Issues        []PatternIssue `json:"issues"`
}

func LoadPatterns(directory string) ([]Pattern, error) {
	directory = ResolvePatternDirectory(directory)
	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, fmt.Errorf("could not read pattern directory at %s: %w", directory, err)
	}
	var files []string
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".pattern.json") {
			files = append(files, filepath.Join(directory, entry.Name()))
		}
	}
	sort.Strings(files)
	patterns := make([]Pattern, 0, len(files))
	for _, file := range files {
		data, err := os.ReadFile(file)
		if err != nil {
			return nil, fmt.Errorf("could not read pattern file %s: %w", file, err)
		}
		var pattern Pattern
		if err := json.Unmarshal(data, &pattern); err != nil {
			return nil, fmt.Errorf("%s: %w", filepath.Base(file), err)
		}
		patterns = append(patterns, pattern)
	}
	return patterns, nil
}

func ResolvePatternDirectory(directory string) string {
	if directory == "" {
		directory = DefaultPatternDirectory
	}
	if directory != DefaultPatternDirectory {
		return directory
	}
	candidates := []string{DefaultPatternDirectory}
	if cwd, err := os.Getwd(); err == nil {
		for dir := cwd; dir != filepath.Dir(dir); dir = filepath.Dir(dir) {
			candidates = append(candidates, filepath.Join(dir, DefaultPatternDirectory))
		}
	}
	if executable, err := os.Executable(); err == nil {
		executableDir := filepath.Dir(executable)
		candidates = append(candidates,
			filepath.Join(executableDir, DefaultPatternDirectory),
			filepath.Join(executableDir, "..", "share", "loom", DefaultPatternDirectory),
		)
	}
	for _, candidate := range candidates {
		if entries, err := os.ReadDir(candidate); err == nil {
			for _, entry := range entries {
				if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".pattern.json") {
					return candidate
				}
			}
		}
	}
	return directory
}

func ValidatePatterns(directory string) PatternValidationReport {
	resolved := ResolvePatternDirectory(directory)
	abs, _ := filepath.Abs(resolved)
	patterns, err := LoadPatterns(resolved)
	if err != nil {
		return PatternValidationReport{"1", "error", abs, 0, []PatternIssue{{SeverityError, "PATTERN001", abs, "pattern directory cannot be read."}}}
	}
	issues := []PatternIssue{}
	if len(patterns) == 0 {
		issues = append(issues, PatternIssue{SeverityError, "PATTERN003", abs, "No .pattern.json files were found."})
	}
	ids := map[string]bool{}
	kinds := map[NodeKind]bool{}
	for _, pattern := range patterns {
		path := pattern.ID + ".pattern.json"
		if pattern.SchemaVersion != "1" {
			issues = append(issues, PatternIssue{SeverityError, "PATTERN004", path, "schema_version must be 1."})
		}
		if pattern.ID == "" || strings.Contains(pattern.ID, "_") {
			issues = append(issues, PatternIssue{SeverityError, "PATTERN005", path, "id must use lowercase letters, numbers, and hyphens."})
		}
		if ids[pattern.ID] {
			issues = append(issues, PatternIssue{SeverityError, "PATTERN007", path, "Duplicate pattern id " + pattern.ID + "."})
		}
		ids[pattern.ID] = true
		if kinds[pattern.Kind] {
			issues = append(issues, PatternIssue{SeverityError, "PATTERN008", path, "Duplicate semantic kind " + string(pattern.Kind) + "."})
		}
		kinds[pattern.Kind] = true
		if pattern.Intent.Summary == "" || pattern.Semantics.Role == "" || pattern.Category == "" {
			issues = append(issues, PatternIssue{SeverityError, "PATTERN009", path, "Intent, semantic role, and category must be non-empty."})
		}
		if len(pattern.Mappings) == 0 {
			issues = append(issues, PatternIssue{SeverityError, "PATTERN015", path, "Mappings must be non-empty."})
		}
		for _, attribute := range pattern.Attributes {
			if attribute.Minimum != nil && attribute.Maximum != nil && *attribute.Minimum > *attribute.Maximum {
				issues = append(issues, PatternIssue{SeverityError, "PATTERN012", path + "#attributes." + attribute.Name, "minimum cannot exceed maximum."})
			}
		}
	}
	status := "ok"
	if len(issues) > 0 {
		status = "error"
	}
	return PatternValidationReport{"1", status, abs, len(patterns), issues}
}

func LintPatterns(directory string) PatternValidationReport {
	report := ValidatePatterns(directory)
	if report.Status != "ok" {
		return report
	}
	patterns, _ := LoadPatterns(directory)
	for _, pattern := range patterns {
		platforms := map[string]bool{}
		for _, mapping := range pattern.Mappings {
			platforms[mapping.Platform] = true
		}
		if !platforms["swiftui"] {
			report.Issues = append(report.Issues, PatternIssue{SeverityError, "PATTERN102", pattern.ID + ".pattern.json", "pattern must include a swiftui mapping."})
		}
		if !platforms["winui3"] {
			report.Issues = append(report.Issues, PatternIssue{SeverityError, "PATTERN103", pattern.ID + ".pattern.json", "pattern must include a winui3 mapping."})
		}
		if !platforms["qt"] {
			report.Issues = append(report.Issues, PatternIssue{SeverityError, "PATTERN104", pattern.ID + ".pattern.json", "pattern must include a qt mapping."})
		}
	}
	if len(report.Issues) > 0 {
		report.Status = "error"
	}
	return report
}

func PatternListText(patterns []Pattern) string {
	width := 0
	for _, pattern := range patterns {
		if len(pattern.ID) > width {
			width = len(pattern.ID)
		}
	}
	var b strings.Builder
	for _, pattern := range patterns {
		fmt.Fprintf(&b, "%-*s  %s  %s\n", width, pattern.ID, pattern.Kind, pattern.Intent.Summary)
	}
	return b.String()
}

func FindPattern(patterns []Pattern, id string) (Pattern, bool) {
	for _, pattern := range patterns {
		if pattern.ID == id {
			return pattern, true
		}
	}
	return Pattern{}, false
}

func ExportPatterns(patterns []Pattern, format string) (string, error) {
	switch format {
	case "", "loom":
		return prettyJSON(patterns)
	case "dtcg":
		type token struct {
			Type        string             `json:"$type"`
			Value       any                `json:"$value"`
			Description string             `json:"$description"`
			Extensions  map[string]Pattern `json:"$extensions"`
		}
		out := map[string]any{"$description": "loom OS-agnostic UI pattern catalog exported as DTCG-compatible token objects.", "loom": map[string]token{}}
		tokens := out["loom"].(map[string]token)
		for _, pattern := range patterns {
			tokens[pattern.ID] = token{"other", map[string]any{"id": pattern.ID, "name": pattern.Name, "kind": pattern.Kind, "category": pattern.Category}, pattern.Intent.Summary, map[string]Pattern{"loom": pattern}}
		}
		return prettyJSON(out)
	case "open-ui":
		return prettyJSON(map[string]any{"schema_version": "1", "source": "loom", "components": patterns})
	case "aria":
		type aria struct {
			ID            string   `json:"id"`
			Name          string   `json:"name"`
			Role          string   `json:"role"`
			NameSource    string   `json:"nameSource"`
			FocusBehavior string   `json:"focusBehavior"`
			Notes         []string `json:"notes"`
		}
		var entries []aria
		for _, pattern := range patterns {
			entries = append(entries, aria{pattern.ID, pattern.Name, pattern.Accessibility.Role, pattern.Accessibility.NameSource, pattern.Accessibility.FocusBehavior, pattern.Accessibility.Notes})
		}
		return prettyJSON(map[string]any{"schema_version": "1", "source": "loom", "patterns": entries})
	case "style-dictionary":
		out := map[string]any{"loom": map[string]any{}}
		items := out["loom"].(map[string]any)
		for _, pattern := range patterns {
			items[pattern.ID] = map[string]any{"value": pattern.ID, "type": "loom.pattern", "comment": pattern.Intent.Summary, "attributes": map[string]any{"name": pattern.Name, "kind": pattern.Kind, "category": pattern.Category}}
		}
		return prettyJSON(out)
	default:
		return "", fmt.Errorf("--format must be loom, dtcg, open-ui, aria, or style-dictionary")
	}
}
